tket/src/Transformations/CliffordOptimisation.cpp

```diff
@@ -699,6 +705,250 @@ Transform singleq_clifford_sweep() {
+          frontier.insert({in_edges[source], qubits[i]});
+          frontier.erase(e);
+        }
+      }
+    }
+  }
+  return std::make_pair(clifford_vertices, end_of_circuit_measures);
+}
+
+Transform push_cliffords_through_measures() {
+  return Transform([](Circuit &circ) {
+    // Extract Clifford and Measure vertices information from provided circuit
+    std::pair<VertexSet, std::vector<MeasureVertices>> clifford_info =
+        get_end_of_circuit_clifford(circ);
+    VertexSet clifford_vertices = clifford_info.first;
+    std::vector<MeasureVertices> measure_vertices = clifford_info.second;
+
+    // Initialize vectors to track various information
+    std::vector<Qubit> qubits;
+    std::vector<Bit> bits;
+    QubitPauliMap base;
+    VertexSet final_measure_vertices;
+
+    // Populate qubits, bits, basic measurement operators and final measurement
+    // vertices
+    for (const MeasureVertices &mv : measure_vertices) {
+      qubits.push_back(mv.qubit);
+      bits.push_back(mv.bit);
+      base.insert({mv.qubit, Pauli::I});
+      final_measure_vertices.insert(mv.measure);
+    }
+
+    // Construct a circuit with Clifford gates
+    Circuit clifford_circuit =
+        circ.subcircuit(circ.make_subcircuit(clifford_vertices));
+    UnitaryRevTableau cliff_tab =
+        circuit_to_unitary_rev_tableau(clifford_circuit);
+
+    // Generate updated measurement operator for each end of circuit measurement
+    std::list<SpSymPauliTensor> measurement_operators;
+    for (const Qubit &q : qubits) {
+      QubitPauliMap copy = base;
+      copy[q] = Pauli::Z;
+      SpPauliStabiliser sps(copy, 2);
+      measurement_operators.push_back(cliff_tab.get_row_product(sps));
+    }
+
+    // Mutually diagonalize updated measuement operators
+    Circuit mutual_c = mutual_diagonalise(
+        measurement_operators, std::set<Qubit>(qubits.begin(), qubits.end()),
+        CXConfigType::Snake);
+
+    // If mutual diagonalisation circuit doesn't improve 2-qubit gate
+    // count then keep circuit as is
+    if (mutual_c.count_n_qubit_gates(2) >=
+        clifford_circuit.count_n_qubit_gates(2)) {
+      return false;
+    }
+
+    // Remove Clifford and Measure vertices from the original circuit
+    // before adding mutualy diagonalisation circuit
+    circ.remove_vertices(
+        clifford_vertices, Circuit::GraphRewiring::Yes,
+        Circuit::VertexDeletion::Yes);
+    circ.remove_vertices(
+        final_measure_vertices, Circuit::GraphRewiring::Yes,
+        Circuit::VertexDeletion::Yes);
+    // add mutual diagonalisation circuit
+    circ.append(mutual_c);
+
+    // add back in end of circuit measurements
+    TKET_ASSERT(qubits.size() == bits.size());
+    for (unsigned i = 0; i < qubits.size(); i++) {
+      circ.add_measure(qubits[i], bits[i]);
+    }
+
+    // Add classical logic to permute output measurements to correct result
+    register_t scratch_r =
+        circ.add_c_register("permutation_scratch", bits.size() + 1);
```

Comment:
Can we get an exception for "register already exists"? For example, running the pass twice. Maybe we can use `permutation_scratch` just as a prefix.

Also this might be a bit annoying, but could we move this register name into UnitID.hpp/cpp?

Reply:
Good idea, I've updated the code toget the permutation_scratch name from `UnitID.hpp` and to call `Circuit::get_next_c_reg_name` to make sure a novel register is used each time.